Makefile

```diff
@@ -4,30 +4,62 @@ all: tinygo
 
 .PHONY: all tinygo build/tinygo test $(LLVM_BUILDDIR) llvm-source clean fmt gen-device gen-device-nrf gen-device-avr
 
-LLVM_COMPONENTS = all-targets analysis asmparser asmprinter bitreader bitwriter codegen core coroutines debuginfodwarf executionengine instrumentation interpreter ipo irreader linker lto mc mcjit objcarcopts option profiledata scalaropts support target
-
 UNAME_S := $(shell uname -s)
 ifeq ($(UNAME_S),Linux)
     START_GROUP = -Wl,--start-group
     END_GROUP = -Wl,--end-group
 endif
 
-CLANG_LIBS = $(START_GROUP) $(abspath $(LLVM_BUILDDIR))/lib/libclang.a -lclangAnalysis -lclangARCMigrate -lclangAST -lclangASTMatchers -lclangBasic -lclangCodeGen -lclangCrossTU -lclangDriver -lclangDynamicASTMatchers -lclangEdit -lclangFormat -lclangFrontend -lclangFrontendTool -lclangHandleCXX -lclangHandleLLVM -lclangIndex -lclangLex -lclangParse -lclangRewrite -lclangRewriteFrontend -lclangSema -lclangSerialization -lclangStaticAnalyzerCheckers -lclangStaticAnalyzerCore -lclangStaticAnalyzerFrontend -lclangTooling -lclangToolingASTDiff -lclangToolingCore -lclangToolingInclusions $(END_GROUP) -lstdc++
-
 LLD_LIBS = $(START_GROUP) -llldCOFF -llldCommon -llldCore -llldDriver -llldELF -llldMachO -llldMinGW -llldReaderWriter -llldWasm -llldYAML $(END_GROUP)
 
+# If we compiled our own LLVM, then use it.
+ifneq ("$(wildcard $(LLVM_BUILDDIR)/bin/clang)","")
+    OUR_CC = $(LLVM_BUILDDIR)/bin/clang
+endif
+ifneq ("$(wildcard $(LLVM_BUILDDIR)/bin/clang++)","")
+    OUR_CXX = $(LLVM_BUILDDIR)/bin/clang++
+endif
+ifneq ("$(wildcard $(LLVM_BUILDDIR)/bin/clang-8)","")
+    OUR_CC = $(LLVM_BUILDDIR)/bin/clang-8
+endif
+ifneq ("$(wildcard $(LLVM_BUILDDIR)/bin/clang++-8)","")
+    OUR_CXX = $(LLVM_BUILDDIR)/bin/clang++-8
+endif
+ifneq ("$(wildcard $(LLVM_BUILDDIR)/bin/llvm-config)","")
+    LLVM_CONFIG = $(LLVM_BUILDDIR)/bin/llvm-config
+endif
 
-# For static linking.
-CGO_CPPFLAGS=$(shell $(LLVM_BUILDDIR)/bin/llvm-config --cppflags) -I$(abspath $(CLANG_SRC))/include -I$(abspath $(LLD_SRC))/include
-CGO_CXXFLAGS=-std=c++11
-CGO_LDFLAGS=-L$(LLVM_BUILDDIR)/lib $(CLANG_LIBS) $(LLD_LIBS) $(shell $(LLVM_BUILDDIR)/bin/llvm-config --ldflags --libs --system-libs $(LLVM_COMPONENTS))
+# If we don't have our own compiled clang & clang++ yet, look for system provided ones.
+ifeq ("$(OUR_CC)","")
+    OUR_CC = $(shell which clang)
+endif
+ifeq ("$(OUR_CC)","")
+    OUR_CC = $(shell which clang-8)
+endif
+ifeq ("$(OUR_CXX)","")
+    OUR_CXX = $(shell which clang++)
+endif
+ifeq ("$(OUR_CXX)","")
+    OUR_CXX = $(shell which clang++-8)
+endif
 
+# Only set the variables if llvm-config was found, otherwise weird errors appear for the user.
+ifneq ("$(wildcard $(LLVM_CONFIG))","")
+    CLANG_LIBS = $(shell $(LLVM_CONFIG) --libs all) $(abspath $(LLVM_BUILDDIR))/lib/libclang.a
+    TMP_LDFLAGS = $(shell $(LLVM_CONFIG) --ldflags)
+    CGO_LIBS = $(shell $(LLVM_CONFIG) --libs all)
+    CGO_SYSLIBS = $(shell $(LLVM_CONFIG) --system-libs all)
```

Comment:
Note - these calls to `llvm-config` are done one-per-line purely as a way to simplify debugging.  eg be able to more easily display/control/change the information for a given option.

Can put them back onto one line if there's a preference for it, but figured they're probably useful left as-is.